client/src/_direct_session.rs

```diff
@@ -0,0 +1,454 @@
+            forwards: Arc::new(std::sync::RwLock::new(Default::default())),
+            forward_pause: Default::default(),
+        }))
+    }
+
+    pub async fn send(
+        &self,
+        target: NodeId,
+        packet: Payload,
+        transport: TransportType,
+        encrypted: bool,
+    ) -> anyhow::Result<()> {
+        let router_id = self.owner.default_id;
+        let slot = if router_id == target {
+            FORWARD_SLOT_ID
+        } else {
+            self.find_slot(&target)
+                .ok_or(SessionError::Internal(format!(
+                    "Session with [{router_id}] doesn't allow to forward packets for [{target}]"
+                )))?
+        };
+
+        let mut forward = match transport {
+            TransportType::Unreliable => Forward::unreliable(self.raw.id, slot, packet),
+            TransportType::Reliable => Forward::new(self.raw.id, slot, packet),
+            TransportType::Transfer => Forward::new(self.raw.id, slot, packet),
+        };
+
+        let size = forward.encoded_len();
+        if encrypted {
+            forward.set_encrypted();
+        }
+
+        self.wait_for_resume().await;
+        self.raw.send(forward).await?;
+
+        self.record_outgoing(target, transport, size);
+        Ok(())
+    }
+
+    pub fn remove_by_slot(&self, id: SlotId) -> anyhow::Result<NodeId> {
+        let mut forwards = self.forwards.write().unwrap();
+        if let Some(node_id) = forwards.remove_by_slot(id) {
+            return Ok(node_id);
+        }
+        bail!(
+            "Slot {id} not found in session: {} ({})",
+            self.raw.id,
+            self.owner.default_id
+        )
+    }
+
+    pub fn remove(&self, node_id: &NodeId) -> anyhow::Result<NodeEntry<NodeId>> {
+        let mut forwards = self.forwards.write().unwrap();
+        match forwards.remove(node_id) {
+            Some(node_id) => Ok(node_id),
+            None => bail!(
+                "Node {node_id} not found in session: {} ({})",
+                self.raw.id,
+                self.owner.default_id
+            ),
+        }
+    }
+
+    pub fn list(&self) -> Vec<NodeEntry<NodeId>> {
+        let forwards = self.forwards.read().unwrap();
+        forwards.slots.values().cloned().collect()
+    }
+
+    pub fn register(&self, node: NodeEntry<NodeId>, slot: SlotId) {
+        let mut forwards = self.forwards.write().unwrap();
+        forwards.add(node, slot)
+    }
+
+    pub fn get_by_slot(&self, slot: SlotId) -> Option<NodeEntry<NodeId>> {
+        let forwards = self.forwards.read().unwrap();
+        forwards.get_by_slot(slot)
+    }
+
+    pub fn get_by_id(&self, node_id: &NodeId) -> Option<NodeEntry<NodeId>> {
+        let forwards = self.forwards.read().unwrap();
+        forwards.get_by_id(node_id)
+    }
+
+    pub fn find_slot(&self, node_id: &NodeId) -> Option<SlotId> {
+        let forwards = self.forwards.read().unwrap();
+        forwards.get_slot(node_id)
+    }
+
+    #[inline]
```

Comment:
I am sceptical about `inline` suggestions.
It does not guarantee that fn is inlined or not. Dit you make any measurements that inline suggestions here help?
More on that topic here: https://nnethercote.github.io/perf-book/inlining.html